marlowe-playground-client/src/HaskellEditor/View.purs

```diff
@@ -0,0 +1,162 @@
+        [ div [ classes (codeEditor $ state ^. _showBottomPanel) ]
+            [ haskellEditor state ]
+        ]
+    , bottomPanel state
+    ]
+
+otherActions :: forall p. State -> HTML p Action
+otherActions state =
+  div [ classes [ ClassName "group" ] ]
+    [ editorOptions state
+    , compileButton state
+    , sendResultButton state "Send To Simulator" SendResultToSimulator
+    , sendResultButton state "Send To Blockly" SendResultToBlockly
+    ]
+
+editorOptions :: forall p. State -> HTML p Action
+editorOptions state =
+  div [ class_ (ClassName "editor-options") ]
+    [ select
+        [ HTML.id_ "editor-options"
+        , class_ (ClassName "dropdown-header")
+        , onSelectedIndexChange (\idx -> ChangeKeyBindings <$> toEnum idx)
+        ]
+        (map keybindingItem (upFromIncluding bottom))
+    ]
+  where
+  keybindingItem item =
+    if state ^. _haskellEditorKeybindings == item then
+      option [ class_ (ClassName "selected-item"), HTML.value (show item) ] [ text $ show item ]
+    else
+      option [ HTML.value (show item) ] [ text $ show item ]
+
+haskellEditor ::
+  forall m.
+  MonadAff m =>
+  State ->
+  ComponentHTML Action ChildSlots m
+haskellEditor state = slot _haskellEditorSlot unit component unit (Just <<< HandleEditorMessage)
+  where
+  setup editor =
+    liftEffect do
+      mContents <- LocalStorage.getItem StaticData.bufferLocalStorageKey
+      let
+        contents = fromMaybe HE.escrow mContents
+      model <- Monaco.getModel editor
+      Monaco.setValue model contents
+
+  component = monacoComponent $ HM.settings setup
+
+bottomPanel :: forall p. State -> HTML p Action
+bottomPanel state =
+  div
+    ( [ classes
+          ( if showingBottomPanel then
+              [ analysisPanel ]
+            else
+              [ analysisPanel, collapsed ]
+          )
+      , bottomPanelHeight showingBottomPanel
+      ]
+    )
+    [ div
+        [ classes [ footerPanelBg, ClassName "flip-x" ] ]
+        [ section [ classes [ ClassName "panel-header", aHorizontal ] ]
+            [ div [ classes [ ClassName "panel-sub-header-main", aHorizontal ] ]
+                [ div [ class_ (ClassName "minimize-icon-container") ]
+                    [ a [ onClick $ const $ Just $ ShowBottomPanel (state ^. _showBottomPanel <<< to not) ]
+                        [ img [ classes (minimizeIcon $ state ^. _showBottomPanel), src closeDrawerArrowIcon, alt "close drawer icon" ] ]
+                    ]
+                ]
+            ]
+        , section
+            [ classes [ ClassName "panel-sub-header", aHorizontal, ClassName "panel-contents" ]
+            ]
+            (resultPane state)
+        ]
+    ]
+  where
+  showingBottomPanel = state ^. _showBottomPanel
+
+compileButton :: forall p. State -> HTML p Action
+compileButton state = button [ onClick $ const $ Just Compile ] [ text (if state ^. _compilationResult <<< to isLoading then "Compiling..." else "Compile") ]
```

Comment:
1. Can we stick some newlines in here, please?

2. These are all equivalent:

``` purescript
if state ^. _compilationResult <<< to isLoading then ...
if state ^. _compilationResult <<< is _Loading then ...
if view (_compilationResult <<< to isLoading) state then ...
if view (_compilationResult <<< is _Loading) state then ...
if is (_compilationResult <<< _Loading) state then ...
if has (_compilationResult <<< _Loading) state then ...
```

Personally I think the last is best, but up to you. (Though I generally _dislike_ using `to f` when there's already an existing lens for it.)

Reply:
`isLoading = is _Loading` so they are not all possible, I've changed to `has (_compilationResult <<< _Loading) state` as that seems to be the one that works. I keep forgetting about `has` and `is`, they are very nice.